drivers/clock_control: stm32wl: Changes around use of  HSE as input clock

erwango opened the pull request:
This change around use of HSE on STM32WL is two fold:
- Make use of https://github.com/zephyrproject-rtos/hal_stm32/pull/111 which fixes a bug in STM32Cube LL API. (This is a temporary patch as official bug has been filled internally. Reference provided)
- Add a dedicated binding for STM32WL HSE clock and implement related changes

Fixes #37260

erwango committed:
dts/bindings: Add binding for STM32WL HSE Clock

STM32WL features a specific HSE clock with dedicated properties.
Add a dedicated binding and update STM32 clock control driver
header to take it into account.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/arm/st: wl: Update compatible for hse clock

Use dedicated HSE clock binding for stm32wl series.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/clock_control: stm32wl: Consider new options for HSE clock

On STM32WL, HSE clock can take 2 specific options:
-hse-tcxo
-hse-div2
Enable support for these options.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
west.yml: hal_stm32: Add fixup on STM32WL Utils LL API

Make use of https://github.com/zephyrproject-rtos/hal_stm32/pull/111

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Getting false positives on the compliance check:
```
-:27: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_TCXO) is normally used as IS_ENABLED(CONFIG_STM32_HSE_TCXO)
#27: FILE: drivers/clock_control/clock_stm32_ll_common.c:499:
+	if (IS_ENABLED(STM32_HSE_TCXO)) {

-:30: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_DIV2) is normally used as IS_ENABLED(CONFIG_STM32_HSE_DIV2)
#30: FILE: drivers/clock_control/clock_stm32_ll_common.c:502:
+	if (IS_ENABLED(STM32_HSE_DIV2)) {

-:48: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_DIV2) is normally used as IS_ENABLED(CONFIG_STM32_HSE_DIV2)
#48: FILE: drivers/clock_control/clock_stm32_ll_common.c:530:
+	if (IS_ENABLED(STM32_HSE_DIV2)) {

-:65: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_TCXO) is normally used as IS_ENABLED(CONFIG_STM32_HSE_TCXO)
#65: FILE: drivers/clock_control/clock_stm32_ll_common.c:558:
+		if (IS_ENABLED(STM32_HSE_TCXO)) {

-:72: WARNING:IS_ENABLED_CONFIG: IS_ENABLED(STM32_HSE_BYPASS) is normally used as IS_ENABLED(CONFIG_STM32_HSE_BYPASS)
#72: FILE: drivers/clock_control/clock_stm32_ll_common.c:563:
+		if (IS_ENABLED(STM32_HSE_BYPASS)) {
```

ABOSTM commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -542,18 +554,18 @@ int stm32_clock_control_init(const struct device *dev)
 			LL_RCC_HSE_EnableTcxo();
-		} else {
-			LL_RCC_HSE_DisableTcxo();
+		}
```
`else { LL_RCC_HSE_DisableTcxo(); }   ` ?  ... similar to Bypass

erwango replied:
Default configuration is TCXO disabled.
Then, the if condition is defined at build time, so there is no way we can enable then disable.

I think that the one to be fixed would be the bypass part.

ABOSTM commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -486,18 +487,24 @@ int stm32_clock_control_init(const struct device *dev)
 #elif STM32_PLL_SRC_HSE
-	int hse_bypass;
 
+#ifndef CONFIG_SOC_SERIES_STM32WLX
```
nit: why not removing the negation and invert #if and #else ?

ABOSTM reviewed (commented):
2 minor remarks

ABOSTM reviewed (approved):
LGTM

mbolivar-nordic reviewed (approved):
Looks good pending merge of zephyrproject-rtos/hal_stm32#111 and update of west.yml to the final SHA

erwango committed:
dts/arm/seeed: lora-e5: Update hse clock configuration

Following change of HSE clock binding, remove hse-bypass
which is no longer an option and replace it with hse-tcxo
which is the configuration closest to previous status.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

cfriedt commented:
@erwango - please rebase - will merge once it's all green

erwango commented:
> @erwango - please rebase - will merge once it's all green

@cfriedt compliance issue is a false positive (as mentioned her: https://github.com/zephyrproject-rtos/zephyr/pull/37286#issuecomment-888375742).
Aside this, everything is green.
